developerportal/apps/ingestion/utils.py

```diff
@@ -0,0 +1,261 @@
+        # Gets YT thumbnail
+        return entry.media_thumbnail[0]["url"]
+
+    elif (
+        entry
+        and hasattr(entry, "media_content")
+        and entry.media_content[0]["medium"] == "image"
+    ):
+        # Gets RSS image
+        return entry.media_content[0]["url"]
+
+    return ""
+
+
+def fetch_external_data(feed_url: str, last_synced: datetime.datetime) -> list:
+    """For the given feed_url, fetch all entries that are timestamped since
+    last_synced and return them as a list of 0...n standardised dictionaries
+    in the format:
+
+    [
+        {
+            "title": <str> - title of the entry
+            "authors": [<str>] - 0...n author names,
+            "url": <str> - URL of the thing we want to link to as external content,
+            "image_url": <str> - URL of any accompanying image,
+            "timestamp: <datetime.datetime> - item's own timestamp
+        },
+        ...
+    ]
+    """
+
+    output = []
+    parsed_data = feedparser.parse(feed_url)
+
+    # DANGER: we can't do this next check with feeds that lack an <updated> node:
+    if "updated" in parsed_data.feed.keys():
+        # Check we have something that's worth parsing
+        feed_last_updated = parse_datetime(parsed_data.feed.updated)
+        if feed_last_updated <= last_synced:
+            logger.info(
+                f"Feed's last updated date ({feed_last_updated}) "
+                f"was not after last_synced ({last_synced})"
+            )
+            return output
+
+    for entry in parsed_data.entries:
+        timestamp = parse_datetime(entry.published)
+        if timestamp <= last_synced:
+            # no point continuing
+            break
+
+        output.append(
+            dict(
+                title=entry.title,
+                authors=[x["name"] for x in entry.authors],
+                url=entry.link,
+                image_url=_get_item_image(entry),
+                timestamp=timestamp,
+            )
+        )
+
+    return output
+
+
+def _get_model(model_name):
+    if model_name == "Video":
+        module = video_models
+    elif model_name == "ExternalArticle":
+        module = externalcontent_models
+    else:
+        raise NotImplementedError()
+
+    return getattr(module, model_name)
+
+
+def ingest_content(type_: str):
+    """For the given type_:
+    * fetch the relevant data from the configured sources
+    * for each item of source data:
+        * create an appropriate ExternalContent page subclass, as a draft
+    * update the last_sync timestamp for each configured source
+    * submit all just-created draft pages for moderation
+    """
+
+    _now = tz_now()
+    model_name = type_
+
+    configs = IngestionConfiguration.objects.filter(integration_type=type_)
+
+    Model = _get_model(model_name)
+
+    ingestion_user = User.objects.get(username=INGESTION_USER_USERNAME)
+
+    for config in configs:
+        with transaction.atomic():
```

Comment:
I wonder if we'd want to have a mechanism in this function to ensure it's not running multiple times for the same types at the same time. But before I indulge those thoughts: what would happen in that case? Would the second transaction just error out because of some constraints or would we have duplicated items?

Reply:
Ah yeah, we do have a Redis-backed lock we could use (we used to use it to avoid two static builds happening at the same time and fighting with each other). I nearly put it in here, but I thought for the initial version it wasn't worth it - if we end up with potential duplicates, these raise ValidationErrors at save() and are gracefully logged and we move on. 

The only thing I'm not 1000% sure on is if _both_ of two identical imports are happening at te same time, and they're both within their own transaction, whether duplicates will be detected because the commit will happen, in theory, after Python has done checking for a ValidationError based on the slugs committed to the DB at the point the transaction started, if you see what I mean. What might then happen is an IntegrityError, because slugs must be unique - and we'll hear that loudly in Sentry.

Reply:
That does make sense and alleviates my fears. Thanks for the elaboration!